### DISCUSSION TITLE
Simple way to get 'buffer local' `eglot-workspace-configuration`

### DISCUSSION THREAD

**Discussion by dalanicolai** [joaotavora/eglot#1279] [Q&A]
While using [org-src-context.el](https://github.com/karthink/org-src-context/blob/master/org-src-context.el), I am trying to set a buffer local value for  `eglot-workspace-configuration`, as I am trying to set the python virtual environment from the `:kernel/:python` src-block header arg. However, it seems extremely painful to use a buffer local value for the  `eglot-workspace-configuration`. Of course, I could just set the value of  `eglot-workspace-configuration` via `setq-default`, but I would like to not touch its 'default value'.

Now to set the local value from a buffer local variable `python-venv`, I came up with the following solutions
```lisp
(setq-default eglot-workspace-configuration 
	      (lambda (args) 
		`(:pylsp (:plugins
			  (:jedi
			   (:environment ,(buffer-local-value 'python-venv
							      (car (eglot--managed-buffers args)))))))))
```
but I was wondering if there isn't a simpler way to achieve this (this might be very handy while creating some org-mode eglot solution). I have read many discussions/issues here, but I could not find any easy solution. Thank you!

**Comment by joaotavora** [comment]
>  However, it seems extremely painful [...] I have read many discussions/issues here, but I could not find any easy solution.

Have you also thought of consulting Eglot's documentation on the matter? https://joaotavora.github.io/eglot/#Advanced-server-configuration If you have, what is not covered there that you wish would be?

**Reply by dalanicolai** [reply]
Thanks for the quick reply!

For sure, I have looked at the documentation. As far as I can find there, the server can only be configured via
1. 'directory local' variables
2. via server initialization options (incl. passing server command-line options)
3. by configuring the default value of `eglot-workspace-configuration` (via setq-default)

I would like to use the value of a buffer local variable for the configuration, but simply setting a buffer local value for `eglot-workspace-configuration` and then using the `eglot-signal-didChangeConfiguration` does not work. The only way I could find to use a local buffer variable, without changing the default value of `eglot-workspace-configuration` for each individual buffer, was by using the code example above (which uses a 'cumbersome' trick to retrieve the buffer local value).

I was wondering if there was an easier way to use a buffer local value in the `eglot-workspace-configuration`. Maybe I looked over something, or maybe there is something I did not understand correctly.

**Reply by joaotavora** [reply]
> I would like to use the value of a buffer local variable for the configuration, but simply setting a buffer local value for eglot-workspace-configuration and then using the eglot-signal-didChangeConfiguration does not work. 

I think you have to pay extra attention to this part in the documentation and read on from there

```
It’s important to recognize that this variable really only makes sense when set directory-locally. 

It usually does not make sense to set it file-locally or in a major-mode hook.
``` 

So, first decide if you want project-specific configuraiton or user-specific configuration (where user==you, of course). You haven't said which one you want and I can't decide for you.

Only the will you know if `eglot-workspace-configuration` is what you want to use.  If you do decide that you want to use it, then, as the manual outlines, you should set it directory-locally.  And there are two ways to set it directory-locally.  Via files or via Elisp snippets in your configuration.  Decide what you want, and then follow the examples.  There aren't examples for every possible configuraiton.

Read the manual slowly and carefully, I think all  answers you want are there.   

> I was wondering if there was an easier way to use a buffer local value in the eglot-workspace-configuration.

Forget about the "buffer local value" of eglot-workspace-configuration.  That's just an implementation detail that needen't concern you.  You either want project-specific configuration or user-specific configuration.  If you want project-specific config, then DIRECTORY-local (NOT buffer-local) is what you want to use.

**Comment by dalanicolai** [comment]
I think I have read the documentation very carefully. The problem is that I would like to use it with the org 'special edit' buffer, it is not part of a project (org allows for configuring the 'virtual environment' via the `:python` header, or when using [emacs-jupyter](https://github.com/emacs-jupyter/jupyter) via the `:kernel` header; a single org file can contain blocks using different virtual environments). So, I think I would like to use 'user-specific' configuration, but the manual only mentions using the 'command-line arg' options or the `:initializationOptions` for that, I seems that with [pylsp](https://github.com/python-lsp/python-lsp-server) neither of these options can be used to set the (jedi) virtual environment. Therefore, I must use the `eglot-workspace-configuration`, but the manual only explains how to use it via `setq-default`. However, I would like to not touch the 'default' value, but just use a 'buffer local' value. The 'hack' in my example code above achieves what I want, but I was wondering if there exists an 'built-in' solution for it (or if it could be provided).

**Comment by joaotavora** [comment]
Daniel Nicolai ***@***.***> writes:

> So, I think I would like to use 'user-specific' configuration,

Good.  Knowing this is a good outcome of this discussion.

> but the manual only mentions using the 'command-line arg' options or
> the :initializationOptions for that, I seems that with pylsp neither
> of these options can be used to set the (jedi) virtual
> environment.

Well, then, _that's_ the problem.

>Therefore, I must use the eglot-workspace-configuration,

True, or you could ask pylsp to fix this problem.

> but the manual only explains how to use it via setq-default.
> I would like to not touch the 'default' value, but just use a 'buffer
> local' value. The 'hack' in my example code

Because using eglot-workspace-configuration, i.w. a WORKSPACE-specific
configuration mechanisms for USER-specific configuration purposes is
_inherently_ a hack.  There's nothing the LSP client can do about that,
it's the LSP server's territory.

Anyway, while it's undeniable setting e-w-c globally is a hack, it is
usually a pretty harmless hack.  Wat's the big problem with setting the
global value of a variable?  What harm do you expect to come from this
setting?  Or is it just the usual programmer OCD?

**Reply by dalanicolai** [reply]
Haha, well I guess it is the usual programmer OCD then. I don't expect much harm indeed, but it just seems an 'inelegant' solution to me, while I thought providing 'buffer local' configuration options is only a simple thing to do. The eglot design makes sense of course, just I guess it was not 'specifically' designed for the 'org babel' workflow. But indeed it (the lack of the 'user-specific' configuration option) could also be regarded as a 'shortcoming' of pylsp.

Anyway, I understand your 'arguments' and I guess they make more sense than the type of solution I was trying to implement. I just wasn't sure before asking here.

Thanks for your quick and helpful answers and for eglot also of course... it's great!.

**Comment by joaotavora** [comment]
Daniel Nicolai ***@***.***> writes:

> Haha, well I guess it is the usual programmer OCD then. I don't expect
> much harm indeed, but it just seems an 'inelegant' solution to me,
> while I thought providing 'buffer local' configuration options is only
> a simple thing to do. The eglot design makes sense of course, just I
> guess it was not 'specifically' designed for the 'org babel' workflow.

Yes, but as always there are other perspectives...

I don't know the details of how Eglot and Org-babel collaborate, but at
a minimum, I would think at some point a (short lived) Python file is
created where Eglot is enabled.  And since I'm reasonably sure that
files tend to be created in directories.  It should theoretically be
possible somehow in the Org-babel infrastructure to specify the
directory-local values of given variables for these directories (even if
short-lived).

Meaning, you can also view this problem as a workspace-specific
configuration but you have no easy idea in advance of what directory it
is.  It becomes an Org-babel/Eglot problem.

> Thanks for your quick and helpful answers and for eglot also of
> course... it's great!.

You're welcome!

**Reply by dalanicolai** [reply]
Thanks for sharing your 'perspective/suggestion', it is very helpful. Indeed, I understand better now how to think about a possible solution.
